### PR TITLE
Release merlin (4.2-411/412 and 3.5.0) and dot-merlin-reader (3.5.0)

### PR DESCRIPTION
**Pull request by voodoos**
Project page: https://github.com/ocaml/merlin

### v4.2-41X:
  + merlin binary
    - external configuration reading:
      + use relative paths to communicate with Dune when possible. This solves
        issues related to symlinks on Unix and improve Windows support (#1271,
        fixes ocaml/merlin#1288)
      + make the `workdir` configuration value when using the
        `dune ocaml-merlin` configuration provider the same as when using
        `dot-merlin-reader` so that ppxes behaves in the same way as before
        (#1284, fixes ocaml/dune#4479, discussion in #1292)
    - destruct:
      + improve prefixing of generated constructors in Destruct by filtering
        opened modules (#1277)
      + make the destruct command more resilient to ill-typed expressions and
        when called without nodes (#1304, fixes ocaml/merlin#1300)
    - reintroduce some record recovery and improve completion (#1276)
    - introduce a new AST node for holes (`_`), allow correct typing of these
      holes and add a new `holes` command that returns the locations of all
      holes in the current file along with their types (#1242, #1289)
    - Mppx: don't restore cookies after invocation. Ppx are invoked only once
      so there is no need to manage cookies. This small change should increase
      performance and should not change any other behavior (#1309)
    - Windows: system command variant: do not open a window console when
      launching a ppx (#1270, fixes ocaml/merlin#714)
    - fix same file documentation bug (#1265 by @ulugbekna, fixes ocaml/merlin#1261)
  + editor modes
    - vim: Add `MerlinNextHole` and `MerlinPreviousHole` commands to navigate
      between holes. Jump to the first hole after destruct (#1287, #1303)
    - emacs: Add `merlin-next-hole` and `merlin-previous-hole` commands to
      navigate holes. Jump to the first hole after calling destruct. (#1291)
    - emacs: modernization of the elisp code and conformance with coding
      guidelines (#1247, #1310 by Steve Purcell )
    - vim & emacs : new client-side "merlin use package" commands, restoring
      previous behavior (#1272, fixes ocaml/merlin#1191)
  + test suite
    - cover constructor disambiguation and record fields (#1276)
    - cover the new `holes` command and AST node (#1242, #1289)
    - cover the document fix (#1265, #1315)

### v3.5.0:
Tue Apr 12 11:44:22 AM CET 2021

  + merlin binary
    - external configuration reading:
      + use relative paths to communicate with Dune when possible. This solves
        issues related to symlinks on Unix and improve Windows support (#1271,
        fixes #1288)
      + make the `workdir` configuration value when using the
        `dune ocaml-merlin` configuration provider the same as when using
        `dot-merlin-reader` so that ppxes behaves in the same way as before
        (#1284, fixes ocaml/dune#4479, discussion in #1292)
    - destruct:  make the destruct command more resilient to ill-typed
      expressions and when called without nodes (#1304, fixes #1300)
    - Mppx: don't restore cookies after invocation. Ppx are invoked only once
      so there is no need to manage cookies. This small change should increase
      performance and should not change any other behavior (#1309)
    - windows:
      + system command variant: do not open a window console when
        launching a ppx (#1270, fixes #714)
      + fix Emacs hanging when starting Merlin (#1263)
      + fix path canonicalization (#1254)
    - fix same file documentation bug (#1265 by @ulugbekna, fixes #1261)
  + editor modes
    - emacs:
      + modernization of the elisp code and conformance with coding
        guidelines (#1247, #1310 by Steve Purcell )
      + use opam var where applicable (#1310)
      + fix "wrong number of argument" (#1250 by @atharvashukla, fixes #1234)
      + fix for Neovim's CursorMoved semantics (#1213 by @ddickstein)
    - vim & emacs : new client-side "merlin use package" commands, restoring
      previous behavior (#1272, fixes #1191)
  + test suite
    - cover the document fix (#1265, #1315)

### CODE REVIEW

**Comment by voodoos** [comment]
Expected CI failures:

- For 3.5.0 only OCamlCI-4.02 should fail, all other tests should pass all the time. 
  (we did not find a solution to prevent OCI from installing test deps and running tests on 4.02 only.)

- For 4.2-41X the random failures on tests `typer-cache`, `file-cache` and `warnings-backtrack` are still an issue (but should be fixed with Dune 2.9)

File: packages/dot-merlin-reader/dot-merlin-reader.3.5.0/opam [hunk 1]
```diff
@@ -0,0 +1,29 @@
+opam-version: "2.0"
+maintainer:   "defree@gmail.com"
+authors:      "The Merlin team"
+synopsis:     "Reads config files for merlin"
+homepage:     "https://github.com/ocaml/merlin"
+bug-reports:  "https://github.com/ocaml/merlin/issues"
+dev-repo: "git+https://github.com/ocaml/merlin.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/dot-merlin-reader/dot-merlin-reader.3.5.0/opam:9]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/merlin/merlin.3.5.0/opam [hunk 2]
```diff
@@ -0,0 +1,77 @@
+opam-version: "2.0"
+maintainer:   "defree@gmail.com"
+authors:      "The Merlin team"
+homepage:     "https://github.com/ocaml/merlin"
+bug-reports:  "https://github.com/ocaml/merlin/issues"
+dev-repo: "git+https://github.com/ocaml/merlin.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/merlin/merlin.3.5.0/opam:8]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/merlin/merlin.4.2-411/opam [hunk 3]
```diff
@@ -0,0 +1,76 @@
+opam-version: "2.0"
+maintainer:   "defree@gmail.com"
+authors:      "The Merlin team"
+homepage:     "https://github.com/ocaml/merlin"
+bug-reports:  "https://github.com/ocaml/merlin/issues"
+dev-repo: "git+https://github.com/ocaml/merlin.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/merlin/merlin.4.2-411/opam:8]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/merlin/merlin.4.2-412/opam [hunk 4]
```diff
@@ -0,0 +1,76 @@
+opam-version: "2.0"
+maintainer:   "defree@gmail.com"
+authors:      "The Merlin team"
+homepage:     "https://github.com/ocaml/merlin"
+bug-reports:  "https://github.com/ocaml/merlin/issues"
+dev-repo: "git+https://github.com/ocaml/merlin.git"
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/merlin/merlin.4.2-412/opam:8]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/dot-merlin-reader/dot-merlin-reader.3.5.0/opam [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  ["dune" "build" "-p" name "-j" jobs]
+]
+depends: [
+  "ocaml" {>= "4.02.1" & < "4.12"}
+  "dune" {>= "1.8.0"}
```
Comment (mseri) [inline packages/dot-merlin-reader/dot-merlin-reader.3.5.0/opam:14]:
```suggestion
  "dune" {>= "2.7.0"}
```
I noticed that upstream seems to be already fixed

**Review by mseri — CHANGES_REQUESTED** [review]

**Comment by mseri** [comment]
Are the tests failures in the extras build related to the sandbox? (See https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/22e8ac056690645795d582ae4497a8acf54c112d)

File: packages/merlin/merlin.4.2-412/opam [hunk 6]
```diff
@@ -0,0 +1,76 @@
+opam-version: "2.0"
+maintainer:   "defree@gmail.com"
+authors:      "The Merlin team"
+homepage:     "https://github.com/ocaml/merlin"
+bug-reports:  "https://github.com/ocaml/merlin/issues"
+dev-repo: "git+https://github.com/ocaml/merlin.git"
+build: [
```
Comment (mseri) [inline packages/merlin/merlin.4.2-412/opam:7]:
```suggestion
license: "MIT"
build: [
```

File: packages/merlin/merlin.4.2-411/opam [hunk 7]
```diff
@@ -0,0 +1,76 @@
+opam-version: "2.0"
+maintainer:   "defree@gmail.com"
+authors:      "The Merlin team"
+homepage:     "https://github.com/ocaml/merlin"
+bug-reports:  "https://github.com/ocaml/merlin/issues"
+dev-repo: "git+https://github.com/ocaml/merlin.git"
+build: [
```
Comment (mseri) [inline packages/merlin/merlin.4.2-411/opam:7]:
```suggestion
license: "MIT"
build: [
```

File: packages/merlin/merlin.3.5.0/opam [hunk 8]
```diff
@@ -0,0 +1,77 @@
+opam-version: "2.0"
+maintainer:   "defree@gmail.com"
+authors:      "The Merlin team"
+homepage:     "https://github.com/ocaml/merlin"
+bug-reports:  "https://github.com/ocaml/merlin/issues"
+dev-repo: "git+https://github.com/ocaml/merlin.git"
+build: [
```
Comment (mseri) [inline packages/merlin/merlin.3.5.0/opam:7]:
```suggestion
license: "MIT"
build: [
```

File: packages/dot-merlin-reader/dot-merlin-reader.3.5.0/opam [hunk 9]
```diff
@@ -0,0 +1,29 @@
+opam-version: "2.0"
+maintainer:   "defree@gmail.com"
+authors:      "The Merlin team"
+synopsis:     "Reads config files for merlin"
+homepage:     "https://github.com/ocaml/merlin"
+bug-reports:  "https://github.com/ocaml/merlin/issues"
+dev-repo: "git+https://github.com/ocaml/merlin.git"
+build: [
```
Comment (mseri) [inline packages/dot-merlin-reader/dot-merlin-reader.3.5.0/opam:8]:
```suggestion
license: "MIT"
build: [
```

**Review by mseri — CHANGES_REQUESTED** [review]
The license field is missing. I have added MIT, which seems the one used in the repo. Can you double-check that this is correct, and upstream this change?

File: packages/dot-merlin-reader/dot-merlin-reader.3.5.0/opam (hunk 5 above)
Comment (voodoos) [reply packages/dot-merlin-reader/dot-merlin-reader.3.5.0/opam]: It is not fixed on every branch, I will change that.

**Comment by voodoos** [comment]
> Are the tests failures in the extras build related to the sandbox? (See https://opam.ci.ocaml.org/github/ocaml/opam-repository/commit/22e8ac056690645795d582ae4497a8acf54c112d)

I believe so, these tests are a bit fragile and I will rework them. I think they should not block the release.
What is exactly the difference between the extra jobs and the other ones ? I did not succeed in reproducing the failure on my machine: I have an unrelated error when copy/pasting the docker rules. Is there another way to reproduce these ?

> The license field is missing. I have added MIT, which seems the one used in the repo. Can you double-check that this is correct, and upstream this change?

Will do, along with the other changes.

**Comment by voodoos** [comment]
Running:
`~/.opam/opam-init/hooks/sandbox.sh build dune runtest -p merlin,dot-merlin-reader`

Pass on my setup...

**Comment by voodoos** [comment]
I was able to reproduce with:
`opam remove -y merlin && opam install -vyt merlin `

**Comment by mseri** [comment]
Super, thanks! The failure on alpine is independent of this PR, so you can ignore it there. I am glad you could reproduce it locally, I agree that they should not make a difference.

I don't know if extra manual testing on BSD is needed in this case, so I wait for @kit-ty-kate and if it's not (or it's done), she will merge

**Comment by kit-ty-kate** [comment]
Thanks a lot!

**Pull request merged**